src/sobol_sensitivity.jl

```diff
@@ -0,0 +1,144 @@
+function give_rand_p(f,p_range,p_fixed=nothing,indices=nothing)
+    if p_fixed == nothing
+        p = [(p_range[j][2] -p_range[j][1])*rand() + p_range[j][1] for j in 1:length(p_range)]
+    else
+        p = [0.0 for i in p_range]
+        j = 1
+        for i in 1:length(p_range)
+            if i in indices
+                p[i] = p_fixed[j]
+                j += 1
+            else
+                p[i] = (p_range[i][2] -p_range[i][1])*rand() + p_range[i][1]
+            end
+        end
+    end
+    p
+end
+
+function calc_mean_var(f,p_range,N)
+    y1 = give_rand_y(f,p_range)
+    y0 = [0.0 for j in 1:length(y1)]
+    v = [0.0 for j in 1:length(y1)]
+    if length(size(y1)) != 1
+        y0 = reshape(y0,size(y1)[1],size(y1)[2])
+        v = reshape(v,size(y1)[1],size(y1)[2])
+    end
+    for i in 1:N
+        y1 = give_rand_y(f,p_range)
+        y0 .+= y1
+        y1 = give_rand_y(f,p_range)
+        v .+= y1.^2
+    end
+    y0 = y0./N
+    y0_sq = [i.^2 for i in y0]
+    v = v./N .- y0_sq
+    y0,v
+end
+
+function first_order_var(f,p_range,N,y0)
+    ys = []
+    for i in 1:length(p_range)
+        y = [0.0 for i in 1:length(y0)]
+        if length(size(y0)) != 1
+            y = reshape(y,size(y0)[1],size(y0)[2])
+        end
+        for j in 1:N
+            p2 = give_rand_p(f,p_range)
+            p1 = give_rand_p(f,p_range,[p2[i]],[i])
+            yer = Array(f(p1)) .* Array(f(p2))
```

Comment:
the mean/var calculations should utilize these same `f` calls. It should just be one function with some options to switch the sampling on/off for second order moments, but you can re-use a lot of computations here.